cpp/src/arrow/compute/kernels/minmax.cc

```diff
@@ -0,0 +1,182 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <algorithm>
+#include <limits>
+#include <utility>
+
+#include "arrow/compute/kernels/aggregate.h"
+#include "arrow/compute/kernels/minmax.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/checked_cast.h"
+
+namespace arrow {
+
+using internal::checked_cast;
+
+namespace compute {
+
+template <typename ArrowType, typename Enable = void>
+struct MinMaxState {};
+
+template <typename ArrowType>
+struct MinMaxState<ArrowType,
+                   typename std::enable_if<is_integer_type<ArrowType>::value>::type> {
+  using ThisType = MinMaxState<ArrowType>;
+  using c_type = typename ArrowType::c_type;
+
+  ThisType& operator+=(const ThisType& rhs) {
+    this->min = std::min(this->min, rhs.min);
+    this->max = std::max(this->max, rhs.max);
+    return *this;
+  }
+
+  void MergeOne(c_type value) {
+    this->min = std::min(this->min, value);
+    this->max = std::max(this->max, value);
+  }
+
+  c_type min = std::numeric_limits<c_type>::max();
+  c_type max = std::numeric_limits<c_type>::min();
+};
+
+template <typename ArrowType>
+struct MinMaxState<ArrowType,
+                   typename std::enable_if<is_floating_type<ArrowType>::value>::type> {
+  using ThisType = MinMaxState<ArrowType>;
+  using c_type = typename ArrowType::c_type;
+
+  ThisType& operator+=(const ThisType& rhs) {
+    this->min = std::fmin(this->min, rhs.min);
+    this->max = std::fmax(this->max, rhs.max);
+    return *this;
+  }
+
+  void MergeOne(c_type value) {
+    this->min = std::fmin(this->min, value);
+    this->max = std::fmax(this->max, value);
+  }
+
+  c_type min = std::numeric_limits<c_type>::infinity();
+  c_type max = -std::numeric_limits<c_type>::infinity();
+};
+
+template <typename ArrowType>
+class MinMaxAggregateFunction final
+    : public AggregateFunctionStaticState<MinMaxState<ArrowType>> {
+ public:
+  using StateType = MinMaxState<ArrowType>;
+
+  explicit MinMaxAggregateFunction(const MinMaxOptions& options) : options_(options) {}
+
+  Status Consume(const Array& array, StateType* state) const override {
+    StateType local;
+
+    internal::BitmapReader reader(array.null_bitmap_data(), array.offset(),
+                                  array.length());
+    const auto values =
+        checked_cast<const typename TypeTraits<ArrowType>::ArrayType&>(array)
+            .raw_values();
+    for (int64_t i = 0; i < array.length(); i++) {
+      if (reader.IsSet()) {
+        local.MergeOne(values[i]);
+      }
+      reader.Next();
+    }
+    *state = local;
+
+    return Status::OK();
+  }
+
+  Status Merge(const StateType& src, StateType* dst) const override {
+    *dst += src;
+    return Status::OK();
+  }
+
+  Status Finalize(const StateType& src, Datum* output) const override {
+    *output = Datum(src.min);
```

Comment:
I think it's fine to return a COLLECTION of the min, max
```suggestion
    *output = Datum({Datum(src.min), Datum(src.max)});
```
But explicitly mention in the doccomment that the output is a COLLECTION of two scalars, the min and max (in that order)

Reply:
Ah, didn't know that a Datum could have a vector of items, thanks! OK, this part turned out to be rather easy to adapt.

Reply:
Should we prefer a struct scalar?

Reply:
Is there a reason to prefer that over COLLECTION? It'll be more work to unpack the struct scalar

Reply:
Scratch that, it's the same. Still, why not COLLECTION?

Reply:
Maybe if we added `StructScalar::GetFieldByName()`:

```c++
Datum out;
MinMax(..., &out);
checked_pointer_cast<StructScalar>(out.scalar())->GetFieldByName("min");
```

Reply:
Struct are easier to evolve since the caller doesn't have to know the absolute index since it's referenced by name. It's also explicit when reading code. We should make a decision global wise for kernel returning complex data.

Reply:
I think since the name of the kernel is "MinMax" either is acceptable in this case